### DISCUSSION TITLE
QQ plot interpretation

### DISCUSSION THREAD

**Discussion by birdwrangler** [bmcclintock/momentuHMM#54] [Q&A] [answered]
I'm modeling accelerometer data (a z value which represents (x + y) as my transmitters do not record a 3rd axis), and attempting to fit a 4 state model with a gamma distribution. About 78% of my data are zeros, so I've specified a zeromass parameter. I currently haven't specified any covariates. I've been using the caracara example as a tutorial. After running multiple iterations with retryFits enabled each time, the Par0 values have stabilized, and AIC suggests the most recent model is best. 

However, I'm not sure what to make of the diagnostic plots and would appreciate any feedback. 
This plot was created from the pseudo residuals using the qqnorm function:
![qqplot](https://user-images.githubusercontent.com/28985900/106976187-fc57de80-6725-11eb-9c0a-936c0f920f69.png)

These plots were created using the function plotPR:
![diagnostics](https://user-images.githubusercontent.com/28985900/106976431-5fe20c00-6726-11eb-827c-ade95ddace11.png)

**Comment by mattgonnerman** [comment]
I just wanted leave a comment on here to add to the discussion. Christine and I had a brief back and forth trying to troubleshoot this as we are both having similar issues with different models. I am modeling a 3 state HMM (roosting, loafing, and foraging). My roost state is primarily based on a high proportion of zero-mass instances in step length (>98%). I have tried varying levels of complexity but the QQPlot always appears to have a flat tail, which I assume is the 0 step lengths.

![QQPlot](https://user-images.githubusercontent.com/46864958/107060159-a5451e80-67a4-11eb-94bc-feddb3d9ece5.PNG)

If I create a QQplot without the 0's in the data, things look much closer to the gamma distribution.
![image](https://user-images.githubusercontent.com/46864958/107060386-e2a9ac00-67a4-11eb-834f-6a040b5e960c.png)

In case it is helpful, here is the code for the model, although I assume the issue relates to the data:
````
nSims <- 100 # number of imputatons
retryFits <- 10 # number attempt to re-fit based on random perturbation
nbStates <- 3 # Number of states
stateNames <- c("roost", "loafing", "foraging") # label states
dist = list(step = "gamma", angle = "wrpcauchy") # distributions for observation processes

## constrain step length parameters: 
# Mean -> dispersal>searching>roosting
# SD -> no relationship
stepDM<-matrix(c(
  1,0,0,0,0,0,0,0,0,
  1,1,0,0,0,0,0,0,0,
  1,1,1,0,0,0,0,0,0,
  0,0,0,1,0,0,0,0,0,
  0,0,0,0,1,0,0,0,0,
  0,0,0,0,0,1,0,0,0,
  0,0,0,0,0,0,1,0,0,
  0,0,0,0,0,0,0,1,0,
  0,0,0,0,0,0,0,0,1),
  nrow = 3*nbStates,byrow=TRUE,
  dimnames=list(c(paste0("mean_",1:nbStates),paste0("sd_",1:nbStates), paste0("zero_",1:nbStates)),
                c("mean_123:(Intercept)", "mean_2","mean_3",
                  paste0("sd_",1:nbStates,":(Intercept)"),
                  paste0("zero_",1:nbStates,":(Intercept)"))))

#define the directions of the differences
stepworkBounds <- matrix(c(-Inf, 0,0,rep(-Inf,6),rep(Inf,9)),nrow = 3*nbStates,
                         dimnames=list(colnames(stepDM),c("lower","upper")))
#Userbound constraint on step
stepBounds <- matrix(c(0,10,
                       0,Inf,
                       0, Inf,
                       0, Inf,
                       0, Inf,
                       0, Inf,
                       .98,1,
                       0,.01,
                       0,.01), nrow = 3*nbStates, byrow = T,
                     dimnames = list(rownames(stepDM), c("lower", "upper")))

## constrain turning angle concentration parameters: 
# Concentration -> searching < dispersal
angleDM<-matrix(c(1,0,0,
                  1,1,0,
                  1,1,1),nrow = nbStates,byrow=TRUE,
                dimnames=list(paste0("concentration_",1:nbStates),
                              c("concentration_1:(Intercept)","concentration_2","concentration_3")))

#Restrict angle concentration such that dispersal > .75 while dispersal>localized
#Userbound contraint on angle
angleBounds <- matrix(c(0,0.94, 
                        0,0.94, 
                        0,0.94),nrow = nbStates,
                      byrow=TRUE, dimnames=list(rownames(angleDM),
                                                c("lower","upper"))) 
#define direction of differences
angleworkBounds <- matrix(c(-Inf,0,0,rep(Inf,3)),
                          nrow = ncol(angleDM), 
                          dimnames=list(colnames(angleDM),c("lower","upper")))


#Bundle individual parameter DM and workbounds
DM<-list(step=stepDM,angle=angleDM)
workBounds<-list(step=stepworkBounds,
                 angle=angleworkBounds)
userBounds <- list(step = stepBounds, 
                   angle = angleBounds)

#prevents working parameters from straying along boundary
prior <- function(par){sum(dnorm(par,0,10,log=TRUE))}

#Fixes "roost" for all, fixPar = fixPar in fitHMM
#fixPar<-list(step=c(rep(NA,nbStates*2),NA, rep(stats::qlogis(1.e-100), 3)))

# initial parameters
Par <- list(step=c(2, 50,120, #mean
                   2, 15,15, #sd
                   .99, .005, .005), #zero mass
            angle = c(.01, 0.2 ,0.3))

Par0_m1.zm <- getParDM(data = turkeyData.zm,
                       nbStates = nbStates, 
                       dist = dist,
                       Par = Par, 
                       DM = DM, 
                       workBounds = workBounds, 
                       userBounds = userBounds,
                       estAngleMean = list(angle = FALSE))

# fit model 1
turk_m1.zm <- fitHMM(data = turkeyData.zm, 
                     nSims = nSims,
                     nbStates = nbStates, 
                     dist = dist, 
                     Par0 = Par0_m1.zm, 
                     DM = DM, 
                     workBounds = workBounds, 
                     userBounds = userBounds,
                     estAngleMean = list(angle=FALSE), 
                     prior = prior,
                     stateNames = stateNames,
                     ncores = 5
)
```

**Comment by bmcclintock** [comment, marked as answer]
As both of you have indicated, the pseudo-residual plots can become quite messy with observed values at the boundary of the support of a data stream (e.g. a value of 0 for a Poisson distribution or a zero-inflated gamma distribution). The Q-Q plot really isn't designed for these cases (or, more generally, for discrete-valued data) because the pseudo-residuals for these observations are the same (hence the flat portions of the line in the Q-Q plot). This can then make the line look like it doesn't fit well for observations away from the boundary. Here's a simple example with a 2-state zero-inflated gamma data stream:
```
set.seed(10,kind="Mersenne-Twister",normal.kind = "Inversion")
ts <- momentuHMM::simData(obsPerAnimal=20000,dist=list(y="gamma"),zeroInflation = list(y=TRUE), Par=list(y=c(10,100,5,10,0.5,1.e-300)))

fit1 <- momentuHMM::fitHMM(ts,nbStates=1,dist=list(y="gamma"),Par0=list(y=c(55,7.5,0.25)))
momentuHMM::plotPR(fit1)
```
![image](https://user-images.githubusercontent.com/11667599/107285941-bf853380-6a14-11eb-8d77-70bd0db1bd58.png)
Now that's a bad fit. But what about with two states:
```
fit2 <- momentuHMM::fitHMM(ts,nbStates=2,dist=list(y="gamma"),Par0=list(y=c(10,100,5,10,0.5,1.e-300)))
momentuHMM::plotPR(fit2)
```
![image](https://user-images.githubusercontent.com/11667599/107286143-083cec80-6a15-11eb-9ce8-bf6b43d6fb82.png)
Despite the zeros in the data messing up the lower quantiles in the Q-Q plot, this is a good fit. But it won't necessarily be this pretty. Here's an example where it's also a good fit, but the flat portions mess up the "slope" for the upper quantiles such that it appears like they don't fit very well (when in fact they do):
```
set.seed(3,kind="Mersenne-Twister",normal.kind = "Inversion")
ts3 <- momentuHMM::simData(obsPerAnimal=20000,dist=list(y="gamma"),zeroInflation = list(y=TRUE), Par=list(y=c(10,100,5,10,0.5,1.e-300)))

fit3 <- momentuHMM::fitHMM(ts3,nbStates=2,dist=list(y="gamma"),Par0=list(y=c(10,100,5,10,0.5,1.e-300)))
momentuHMM::plotPR(fit2)
```
![image](https://user-images.githubusercontent.com/11667599/107286731-d4ae9200-6a15-11eb-9fc2-b6ae4d3c92a9.png)

So the long and the short is that values near the boundary of the support for a data stream can make these plots look funky even when the fit is reasonably good.  While this appears to be the case for @mattgonnerman's model, there seem to be some genuine goodness of fit issues with @birdwrangler's model. The curvature in the upper quantiles of the Q-Q plot indicate some lack of fit, and the ACF plot suggests there is some serious autocorrelation in the data that is not being captured in the model. Additional model structure such as additional states and/or data stream covariates may be needed to capture this.

**Reply by birdwrangler** [reply]
Thanks for this explanation Brett, this helps a lot!
